todo-list-backend/src/index.ts

```diff
@@ -1,33 +1,208 @@
+            .send("request body undefined.  sending response code 400")
+            ; console.log("request body undefined.  sending response code 400"); return
+    }
+
+    if (req.body.id == undefined) {
+        res
+            .status(400)
+            .send("task id undefined.  sending response code 400")
+            ; console.log("task id undefined.  sending response code 400"); return
+    }
+
+    let id = req.body.id
+
+    console.log("removing task id:: " + id)
+
+    let idx = todoItems.findIndex(function (item) { return item.id == id })
+
+    if (idx == -1) {
+        res
+            .status(400)
+            .send("no element in list of requested index")
+            ; console.log("no element in list of requested index"); return
+    }
+
+    let removedElement = todoItems.splice(idx, 1)[0]
+
+    res
+        .status(200)
+        .send(JSON.stringify(removedElement))
+});
+
+app.post("/tag-task-as-complete", express.json(), (req: Request, res: Response) => {
+    console.log("tagging task as complete")
+    const request = req.body;
+    const id = request.id
+    const strict = request.strict;
+    let result
+    try {
+        result = changeCompletionStatusHandler(id, strict, true)
+    } catch (e) {
+        if (e instanceof Error) {
+            res
+                .status(400)
+                .send(e.message)
+                ;
+        } else {
+            res
+                .status(500)
+                .send("unknown backend error")
+        }
+        return
+    }
+    res
+        .setHeader("Access-Control-Allow-Origin", "*")
+        .status(200)
+        .send(JSON.stringify(result))
+    console.log('task tagged as complete')
+});
+
+app.post("/tag-task-as-incomplete", express.json(), (req: Request, res: Response) => {
+    console.log("tagging task as incomplete")
+    const request = req.body;
+    const id = request.id
+    const strict = request.strict;
+    let result
+    try {
+        result = changeCompletionStatusHandler(id, strict, false)
+    } catch (e) {
+        if (e instanceof Error) {
+            res
+                .status(400)
+                .send(e.message)
+                ;
+        } else {
+            res
+                .status(500)
+                .send("unknown backend error")
+        }
+        return
+    }
+    res
+        .setHeader("Access-Control-Allow-Origin", "*")
+        .status(200)
+        .send(JSON.stringify(result))
+});
+
+function changeCompletionStatusHandler(id: number, strict: boolean, newStatus: boolean) {
+    console.log('changeing status of:: '+id+" ::to:: "+newStatus)
+    let idx = todoItems.findIndex(function (value) { return value.id == id })
+
+    if (idx == -1) { throw new Error("no such element") }
+    if (strict && !(Number(newStatus) ^ Number())) { throw new Error("element is already in desired state") }
+
+    todoItems[idx].isDone = newStatus
+
+    saveTodoList(todoItems)
+
+    return todoItems[idx]
+}
+
```

Comment:
This function is done very well, and it makes a lot of sense to move this functionality out into a function for your program, since you're using it in multiple places!